resolver/src/resolver_handle/background.rs

```diff
@@ -0,0 +1,224 @@
+use std::sync::{Arc, Mutex};
+
+use futures::{
+    future,
+    sync::mpsc,
+    Async, Future, Poll, Stream
+};
+use trust_dns_proto::{
+    error::ProtoResult,
+    rr::{Name, RData, RecordType},
+    xfer::{ DnsRequestOptions, RetryDnsHandle,},
+};
+#[cfg(feature = "dnssec")]
+use trust_dns_proto::SecureDnsHandle;
+
+use config::{ResolverConfig, ResolverOpts};
+use dns_lru::DnsLru;
+use hosts::Hosts;
+use lookup::{Lookup, LookupEither, LookupFuture};
+use lookup_ip::LookupIpFuture;
+use lookup_state::CachingClient;
+use name_server_pool::{NameServerPool, StandardConnection};
+
+use super::{BasicResolverHandle, Request};
+
+/// Returns a new future that will drive the background task.
+pub(super) fn task(
+    config: ResolverConfig,
+    options: ResolverOpts,
+    lru: Arc<Mutex<DnsLru>>,
+    request_rx: mpsc::UnboundedReceiver<Request>,
+) -> impl Future<Item = (), Error = ()> {
+    future::lazy(move || {
+        let pool = NameServerPool::<BasicResolverHandle, StandardConnection>::from_config(
```

Comment:
> [...] apologies if the docs and code are clear, I just find it easier to ask the question than assume I read the code correctly [...]

That's fine, I'm happy to answer questions! I think I probably should have explained how this works better in the PR description, and this pattern is fairly complex anyway...

The  `Task` struct, which represents the background work, implements `Future<Item=(), Error=()>` and is constructed and spawned on the executor once per `ResolverHandle`. Once spawned, it continues to run on the executor until all the `ResolverHandle`(or `ResolverHandle`s, as that type is `Clone`) connected to that background task have been dropped, at which point it will return `Async::Ready` the next time it is polled.

The `Task` future is used every time a `ResolverHandle` to it makes a query, however, all of its state (i.e. the `NameServerPool`, etc) is persisted across all queries rather than being reconstructed. When the background future is polled, it checks if there are any incoming resolve requests from its `ResolverHandle`(s), and then returns `Async::NotReady` again once all those requests have been satisfied. If there are no requests, it yields immediately.

The future constructed here in `future::lazy` represents the initial construction of the state to run the background task. Once it finishes, it results in a `Task`, which _also_ implements `Future`, and will continue running until all the handles have been dropped, as described above. This is just so that we don't have to do that work immediately in `ResolverHandle::new` and can wait until the background task has been spawned to set it up, prior to serving any resolve requests.

Does that help clear things up? Please let me know if you have any additional questions!

Reply:
(it occurs to me that some of the description in that comment should probably be recorded more permanently as comments in the code, as well!)

Reply:
Great explanation! Yes this is a pattern I use inside the Proto crate but never extended to the Resolver (though not in such clean implementations).